File: String/3. Longest Substring Without Repeating Characters/Solution.java
/*
Runtime: 4 ms, faster than 87.25% of Java online submissions for Longest Substring Without Repeating Characters.
Memory Usage: 39.1 MB, less than 59.66% of Java online submissions for Longest Substring Without Repeating Characters.
problem solver: gordon fang


Task description
Implement function ToLowerCase() that has a string parameter str, and returns the same string in lowercase.

Example 1:

Input: "Hello"
Output: "hello"
Example 2:

Input: "here"
Output: "here"
Example 3:

Input: "LOVELY"
Output: "lovely"
*/
import java.lang.Math;
import java.util.Set;
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int left = 0;
        Set<Character> set = new HashSet<>();
        for(int right = 0; right< s.length(); right++) {
            char sign = s.charAt(right);
            
            if(!set.add(sign)) {
                while(s.charAt(left) != sign) {
                    set.remove(s.charAt(left++));
                }
                left++;
            }
           
            max = Math.max(max, right - left + 1);
        }
        return max;
    }
}

File: String/3. Longest Substring Without Repeating Characters/Solution.c
/*
Runtime: 4 ms, faster than 82.99% of C online submissions for Longest Substring Without Repeating Characters.
Memory Usage: 5.9 MB, less than 71.01% of C online submissions for Longest Substring Without Repeating Characters.
problem solver: E=EE

Given a string s, find the length of the longest substring without repeating characters.

 

Example 1:

Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
Example 2:

Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
Example 4:

Input: s = ""
Output: 0
*/
int lengthOfLongestSubstring(char * s){
    char* p = s;
    int max = 0;
    int counter[256] = {0};
    while(*s != 0) {
        counter[(char)*s]++;
        if(counter[(char)*s] > 1) {
            while(p<s  && counter[(char)*s]>1) {
                counter[(char)*p]--;
                p++;
            }
        }
        s++;
        max = s-p > max ? s-p : max;
    }
    return max;
}